opensearch: init at 2.5.0

shyim opened the pull request:
###### Description of changes

I maintain the PHP client for Opensearch and packaged it myself as I tested the newest release. Also relates to packaging request #151464

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [X] aarch64-linux
  - [X] x86_64-darwin
  - [X] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Kranzes commented on pkgs/servers/search/opensearch/default.nix:
```diff
@@ -0,0 +1,44 @@
+, zlib
+}:
+
+stdenv.mkDerivation (rec {
+  pname = "opensearch";
+  version = "2.0.0";
+
+  src = fetchurl {
+    url = "https://artifacts.opensearch.org/releases/bundle/opensearch/2.0.0/opensearch-2.0.0-linux-x64.tar.gz";
+    sha256 = "0kjnn1wv46wrgk4m9h6zycgg4n0hiv976s8fxmibjb054790dixy";
+  };
```
```suggestion
stdenv.mkDerivation (rec {
  pname = "opensearch";
  version = "2.1.0";

  src = fetchurl {
    url = "https://artifacts.opensearch.org/releases/bundle/opensearch/${version}/opensearch-${version}-linux-x64.tar.gz";
    sha256 = "sha256-vM1zfg9ORVTTkY8BGh43nr2p5TRzxwtCZKooznGWEeM=";
  };
```

Kranzes replied:
Updated to latest version.

ajs124 commented on pkgs/servers/search/opensearch/default.nix:
```diff
@@ -0,0 +1,44 @@
+  meta = {
+    description = "Open Source, Distributed, RESTful Search Engine";
+    license = lib.licenses.asl20;
+    platforms = lib.platforms.unix;
```
since this fetches a `linux-x64.tar.gz`, isn't this architecture specific?

shyim replied:
I am working on macbook.. Deleting the jre in zip and using own works good :)

shyim commented:
@Kranzes I adopted the most, but I would not restrict the packaging only for the official provided packages. Officially, is only Linux amd64/arm64 packaged. The difference is the packaged Java. 

As we use our own Java, we can support any platform our own java supports 😅

drupol reviewed (approved):
LGTM

fballiano commented:
it seems one of the workflow hung and won't run, can we do something about it?

drupol commented:
@GrahamcOfBorg build opensearch
@GrahamcOfBorg test opensearch

drupol commented:
Maybe @SuperSandro2000 may help in here to push the frustration away.

mweinelt commented:
@ofborg eval

mweinelt commented on pkgs/servers/search/opensearch/default.nix:
```diff
@@ -0,0 +1,46 @@
+    wrapProgram $out/bin/opensearch-plugin --set JAVA_HOME "${jre_headless}"
+  '';
+
+  meta = {
+    description = "Open Source, Distributed, RESTful Search Engine";
+    license = lib.licenses.asl20;
+    platforms = lib.platforms.unix;
+    maintainers = with lib.maintainers; [ shyim ];
+  };
```
Can we get a homepage link in here? Ideally a link to their source code repository, where we can submit issues and browse the code. Should be https://github.com/opensearch-project/OpenSearch.

mweinelt commented on pkgs/servers/search/opensearch/default.nix:
```diff
@@ -0,0 +1,46 @@
+
+  src = fetchurl {
+    url = "https://artifacts.opensearch.org/releases/bundle/opensearch/2.3.0/opensearch-2.3.0-linux-x64.tar.gz";
+    sha256 = "1jhhz6f58iz5l920avnp0aaai3lfqpqwymk447bjlbvd2arh0rb9";
```
```suggestion
    hash = "";
```

Can we migrate this to an SRI hash instead? If you build this using an empty hash you should get one.

mweinelt commented on pkgs/servers/search/opensearch/default.nix:
```diff
@@ -0,0 +1,46 @@
+
+stdenvNoCC.mkDerivation rec {
+  pname = "opensearch";
+  version = "2.3.0";
+
+  src = fetchurl {
+    url = "https://artifacts.opensearch.org/releases/bundle/opensearch/2.3.0/opensearch-2.3.0-linux-x64.tar.gz";
```
```suggestion
  version = "2.3.0";

  src = fetchurl {
    url = "https://artifacts.opensearch.org/releases/bundle/opensearch/${version}/opensearch-${version}-linux-x64.tar.gz";
```

Tie the source URL to the version, so we don't have to modify it on every update. That being said, the latest release seems to be 2.5.0.

And what about other targets than x86_64-linux, are there any?

shyim replied:
We use the JDK from nix. So it makes no difference for any platform. I have it running on MacOS :)

fballiano commented on pkgs/servers/search/opensearch/default.nix:
```diff
@@ -0,0 +1,46 @@
+  '';
+
+  meta = {
+    description = "Open Source, Distributed, RESTful Search Engine";
```
```suggestion
    description = "Open Source, Distributed, RESTful Search Engine";
    homepage = "https://github.com/opensearch-project/OpenSearch";
```

shyim commented:
Applied all comments and tested locally on my Mac
![image](https://user-images.githubusercontent.com/6224096/215839357-79325283-82a1-40eb-ad5d-78d12c91973f.png)

drupol commented on pkgs/servers/search/opensearch/default.nix:
```diff
@@ -0,0 +1,47 @@
+  version = "2.5.0";
+
+  src = fetchurl {
+    url = "https://artifacts.opensearch.org/releases/bundle/opensearch/2.5.0/opensearch-2.5.0-linux-x64.tar.gz";
```
```suggestion
    url = "https://artifacts.opensearch.org/releases/bundle/opensearch/${version}/opensearch-${version}-linux-x64.tar.gz";
```

mweinelt replied:
That one still applies :)

shyim replied:
Added it :)

fballiano commented on pkgs/servers/search/opensearch/default.nix:
```diff
@@ -0,0 +1,47 @@
+  version = "2.5.0";
+
+  src = fetchurl {
+    url = "https://artifacts.opensearch.org/releases/bundle/opensearch/${version}/opensearch-2.5.0-linux-x64.tar.gz";
```
```suggestion
    url = "https://artifacts.opensearch.org/releases/bundle/opensearch/${version}/opensearch-${version}-linux-x64.tar.gz";
```

shyim commented on pkgs/servers/search/opensearch/default.nix:
```diff
@@ -0,0 +1,47 @@
+  version = "2.5.0";
+
+  src = fetchurl {
+    url = "https://artifacts.opensearch.org/releases/bundle/opensearch/${version}/opensearch-2.5.0-linux-x64.tar.gz";
```
haha did it now too 😂

mweinelt commented on pkgs/servers/search/opensearch/default.nix:
```diff
@@ -0,0 +1,47 @@
+  buildInputs = [ jre_headless util-linux ];
+  patches = [./opensearch-home-fix.patch ];
+
+  installPhase = ''
+    mkdir -p $out
+    cp -R bin config lib modules plugins $out
+
+    substituteInPlace $out/bin/opensearch \
+      --replace 'bin/opensearch-keystore' "$out/bin/opensearch-keystore"
+
+    wrapProgram $out/bin/opensearch \
+      --prefix PATH : "${lib.makeBinPath [ util-linux gnugrep coreutils ]}" \
+      --set JAVA_HOME "${jre_headless}"
+
+    wrapProgram $out/bin/opensearch-plugin --set JAVA_HOME "${jre_headless}"
+  '';
```
```suggestion
  installPhase = ''
    runHook preInstall  

    mkdir -p $out
    cp -R bin config lib modules plugins $out

    substituteInPlace $out/bin/opensearch \
      --replace 'bin/opensearch-keystore' "$out/bin/opensearch-keystore"

    wrapProgram $out/bin/opensearch \
      --prefix PATH : "${lib.makeBinPath [ util-linux gnugrep coreutils ]}" \
      --set JAVA_HOME "${jre_headless}"

    wrapProgram $out/bin/opensearch-plugin --set JAVA_HOME "${jre_headless}"

    runHook preInstall
  '';
```

And let's follow the best practice here, to make overriding more predictable.

shyim replied:
done :)

mweinelt commented on pkgs/servers/search/opensearch/default.nix:
```diff
@@ -0,0 +1,51 @@
+
+    wrapProgram $out/bin/opensearch-plugin --set JAVA_HOME "${jre_headless}"
+
+    runHook preInstall
```
```suggestion
    runHook postInstall
```

That should be the last fixup for now.

shyim replied:
done :)

drupol commented on pkgs/servers/search/opensearch/default.nix:
```diff
@@ -0,0 +1,51 @@
+
+    wrapProgram $out/bin/opensearch-plugin --set JAVA_HOME "${jre_headless}"
+
+    runHook preInstall
```
```suggestion
    runHook postInstall
```

mweinelt commented on pkgs/servers/search/opensearch/opensearch-home-fix.patch:
```diff
@@ -0,0 +1,26 @@
+-done
+-OPENSEARCH_HOME=`dirname "$OPENSEARCH_HOME"`
++if [ -z "$OPENSEARCH_HOME" ]; then
++    echo "You must set the OPENSEARCH_HOME var" >&2
```
```
❯ OPENSEARCH_HOME=/tmp/opensearch ./result/bin/opensearch-node
could not find java in bundled jdk at /tmp/opensearch/jdk/bin/java
```

So what it actually requires is essentially like a `JAVA_HOME`? But I cannot just pass in for example `openjdk.out`, because it has `bin/java`, but not `jdk/bin/java`.

```
❯ OPENSEARCH_HOME=/nix/store/ch095in4hr481iqv11s55ks9ksl2c49i-openjdk-19.0.1+10/ ./result/bin/opensearch-node
could not find java in bundled jdk at /nix/store/ch095in4hr481iqv11s55ks9ksl2c49i-openjdk-19.0.1+10//jdk/bin/java
```

So then I symlink myself something together, and now I'm stuck here:

````
❯ OPENSEARCH_HOME=/tmp/opensearch ./result/bin/opensearch
Error: Could not find or load main class org.opensearch.tools.java_version_checker.JavaVersionChecker
Caused by: java.lang.ClassNotFoundException: org.opensearch.tools.java_version_checker.JavaVersionChecker
````

What does it take to get this running? I think we really need a small module to encapsulate this requirement.

shyim replied:
see this as example how to get it up running :)
https://github.com/cachix/devenv/blob/main/src/modules/services/elasticsearch.nix#L35-L80

mweinelt replied:
If they haven't diverge enough already, a compatible module would be amazing.

shyim replied:
Make the existing elastic search one compatible or a new one?

mweinelt replied:
That really depends on how compatible they still are.

shyim replied:
I added a module `services.opensearch.enable = true` and works

![image](https://user-images.githubusercontent.com/6224096/215871251-866fd6ef-024e-4744-b821-fea8ea5f41b1.png)

mweinelt replied:
Cool, what do you think about adding a small nixos test to continuously validate that everything is still working?

shyim replied:
I will look tomorrow at the moment I don't have a machine with kvm enabled

shyim replied:
test is added :)

mweinelt commented on nixos/modules/module-list.nix:
```diff
@@ -1042,6 +1042,7 @@
   ./services/search/elasticsearch.nix
   ./services/search/hound.nix
   ./services/search/kibana.nix
+  ./services/search/opensearch.nix
   ./services/search/meilisearch.nix
```
```suggestion
  ./services/search/meilisearch.nix
  ./services/search/opensearch.nix
```

Proper ordering.

mweinelt commented on nixos/modules/services/search/opensearch.nix:
```diff
@@ -0,0 +1,221 @@
+let
+  cfg = config.services.opensearch;
+
+  opensearchConfig = ''
+    network.host: ${cfg.listenAddress}
+    cluster.name: ${cfg.cluster_name}
+    ${lib.optionalString cfg.single_node "discovery.type: single-node"}
+
+    http.port: ${toString cfg.port}
+    transport.port: ${toString cfg.tcp_port}
+
+    ${cfg.extraConf}
+  '';
+
+  configDir = cfg.dataDir + "/config";
+
+  opensearchYml = pkgs.writeTextFile {
+    name = "opensearch.yml";
+    text = opensearchConfig;
+  };
```
Instead of build the config by hand, try `pkgs.formats.yaml` from RFC 42.

https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md

mweinelt commented on nixos/modules/services/search/opensearch.nix:
```diff
@@ -0,0 +1,221 @@
+      type = types.package;
+    };
+
+    listenAddress = mkOption {
+      description = lib.mdDoc "OpenSearch listen address.";
+      default = "127.0.0.1";
+      type = types.str;
+    };
+
+    port = mkOption {
+      description = lib.mdDoc "OpenSearch port to listen for HTTP traffic.";
+      default = 9200;
+      type = types.port;
+    };
+
+    tcp_port = mkOption {
+      description = lib.mdDoc "OpenSearch port for the node to node communication.";
+      default = 9300;
+      type = types.int;
+    };
+
+    cluster_name = mkOption {
+      description = lib.mdDoc "OpenSearch name that identifies your cluster for auto-discovery.";
+      default = "opensearch";
+      type = types.str;
+    };
```
Put those below `services.opensearch.settings`, so you can later use `format.generate "opensearch.yml cfg.settings`. The options should then use the proper path that the config expects.

shyim replied:
So we leave cluster name and so. Just set inside enable them into `settings` and use `settings` to generate that file

mweinelt replied:
Whatever goes into the config files should be below `settings`.

shyim replied:
Alright. I will look after Fosdem :)

drupol replied:
Hey you're going to Fosdem ! I'll be there too :) I'll be sticking the Nix room on Sunday afternoon hit me up !

shyim replied:
@mweinelt done :)

mweinelt reviewed (changes requested):
Thanks, that looks pretty good already. Two issues:

- The module should use the `settings` option and the formatter from RFC42
- The tests should be in its own commit `nixos/tests/opensearch: init`.

Check the botamusique module for an example on how those bits work.

SuperSandro2000 commented on nixos/tests/opensearch.nix:
```diff
@@ -0,0 +1,19 @@
+  };
+
+  testScript = ''
+        machine.start()
+        machine.wait_for_unit("opensearch.service")
+        machine.wait_for_open_port(9200)
+
+        machine.succeed(
+            "curl --fail localhost:9200"
+        )
+      '';
```
```suggestion
    machine.start()
    machine.wait_for_unit("opensearch.service")
    machine.wait_for_open_port(9200)

    machine.succeed(
        "curl --fail localhost:9200"
    )
  '';
```

SuperSandro2000 commented on nixos/modules/services/search/opensearch.nix:
```diff
@@ -0,0 +1,221 @@
+  };
+in
+{
+
+  ###### interface
+
```
```suggestion
```
Those are legacy comments which get copy pasted. We can remove them.

SuperSandro2000 commented on nixos/modules/services/search/opensearch.nix:
```diff
@@ -0,0 +1,221 @@
+  ###### interface
+
+  options.services.opensearch = {
+    enable = lib.mkOption {
+      description = lib.mdDoc "Whether to enable OpenSearch.";
+      default = false;
+      type = lib.types.bool;
+    };
```
This should use mkEnableOption

SuperSandro2000 commented on nixos/modules/services/search/opensearch.nix:
```diff
@@ -0,0 +1,221 @@
+      type = lib.types.bool;
+    };
+
+    package = lib.mkOption {
+      description = lib.mdDoc "OpenSearch package to use.";
+      default = pkgs.opensearch;
+      defaultText = literalExpression "pkgs.opensearch";
+      type = lib.types.package;
+    };
```
This should be using mkPackageOption

SuperSandro2000 commented on nixos/modules/services/search/opensearch.nix:
```diff
@@ -0,0 +1,221 @@
+      default = true;
+    };
+  };
+
+  ###### implementation
```
nit

```suggestion
```

SuperSandro2000 commented on nixos/modules/services/search/opensearch.nix:
```diff
@@ -0,0 +1,221 @@
+        Restart = "always";
+        TimeoutStartSec = "infinity";
+      };
+      preStart = ''
+        ${optionalString (!config.boot.isContainer) ''
+          # Only set vm.max_map_count if lower than ES required minimum
+          # This avoids conflict if configured via boot.kernel.sysctl
+          if [ `${pkgs.procps}/bin/sysctl -n vm.max_map_count` -lt 262144 ]; then
+            ${pkgs.procps}/bin/sysctl -w vm.max_map_count=262144
+          fi
+        ''}
+
```
```suggestion
      preStart = optionalString (!config.boot.isContainer) ''
        # Only set vm.max_map_count if lower than ES required minimum
        # This avoids conflict if configured via boot.kernel.sysctl
        if [ $(${pkgs.procps}/bin/sysctl -n vm.max_map_count) -lt 262144 ]; then
          ${pkgs.procps}/bin/sysctl -w vm.max_map_count=262144
        fi
     '' + ''
```
we don't need ot inline this here. Also $() is usually used instead of backticks.

SuperSandro2000 commented on nixos/modules/services/search/opensearch.nix:
```diff
@@ -0,0 +1,221 @@
+        ln -sfT ${cfg.package}/lib ${cfg.dataDir}/lib
+        ln -sfT ${cfg.package}/modules ${cfg.dataDir}/modules
+
+        # opensearch needs to create the opensearch.keystore in the config directory
+        # so this directory needs to be writable.
+        mkdir -m 0700 -p ${configDir}
```
Normally we use tmpfiles for directory creation

mweinelt replied:
I wouldn't insist on using tmpfiles, they're a bit on the painful end of things.

SuperSandro2000 commented on nixos/modules/services/search/opensearch.nix:
```diff
@@ -0,0 +1,221 @@
+        mkdir -m 0700 -p ${cfg.dataDir}
+
+        # Install plugins
+        ln -sfT ${cfg.package}/lib ${cfg.dataDir}/lib
+        ln -sfT ${cfg.package}/modules ${cfg.dataDir}/modules
```
For symlinks we can also use tmpfiles

shyim replied:
Do you have an example for me? like other service as reference what you mean :)

SuperSandro2000 replied:
https://github.com/NixOS/nixpkgs/blob/35d1efaf52d8e3956a9a91ed5972fbc2ad1c1f48/nixos/modules/services/cluster/hadoop/yarn.nix#L144-L146

also see ``man tmpfiles.d`` for all the details

talyz replied:
tmpfiles should not be used for this! It's much better to do it as it's already done, as that makes sure the symlinks are set up when the service starts. Doing the same thing with tmpfiles can result in the service being run before the symlinks are created.

SuperSandro2000 commented on nixos/modules/services/search/opensearch.nix:
```diff
@@ -0,0 +1,221 @@
+        # access denied ("java.io.FilePermission" "/var/lib/opensearch/config/opensearch.yml" "read")
+
+        cp ${opensearchYml} ${configDir}/opensearch.yml
+        cat ${configDir}/opensearch.yml
```
```suggestion
```
looks like debugging code to me. This could also leak secrets into logs.

shyim replied:
AHH yes. It was for debugging. I removed it

SuperSandro2000 commented on nixos/modules/services/search/opensearch.nix:
```diff
@@ -0,0 +1,221 @@
+        cp ${cfg.package}/config/jvm.options ${configDir}/jvm.options
+        # redirect jvm logs to the data directory
+        mkdir -m 0700 -p ${cfg.dataDir}/logs
+        ${pkgs.sd}/bin/sd 'logs/gc.log' '${cfg.dataDir}/logs/gc.log' ${configDir}/jvm.options \
```
```suggestion
        sed -e '#logs/gc.log#${cfg.dataDir}/logs/gc.log#' -i ${configDir}/jvm.options \
```
no need to draw in an extra dependency

SuperSandro2000 commented on nixos/modules/services/search/opensearch.nix:
```diff
@@ -0,0 +1,221 @@
+        mkdir -m 0700 -p ${cfg.dataDir}/logs
+        ${pkgs.sd}/bin/sd 'logs/gc.log' '${cfg.dataDir}/logs/gc.log' ${configDir}/jvm.options \
+
+        if [ "$(id -u)" = 0 ]; then chown -R opensearch:opensearch ${cfg.dataDir}; fi
```
preStart is always run as root

mweinelt replied:
Is it? Do you have a reference for that?

shyim replied:
Seems so. I have tested it and adjusted it

SuperSandro2000 replied:
I might have mixed up ExecStartPre with preStart from our start script.

Also that is not right by default. I think the fact I based this knowledge on was using PermissionsStartOnly=true which is off by default. Sorry about that.

SuperSandro2000 commented on pkgs/servers/search/opensearch/default.nix:
```diff
@@ -0,0 +1,51 @@
+      --replace 'bin/opensearch-keystore' "$out/bin/opensearch-keystore"
+
+    wrapProgram $out/bin/opensearch \
+      --prefix PATH : "${lib.makeBinPath [ util-linux gnugrep coreutils ]}" \
```
We don't really need to wrap gnugrep and coreutils since they are always there on nixos.

shyim replied:
I use that package also on macOs. I guess it makes sense to leave it still here or? 🤔

SuperSandro2000 reviewed (commented):
just some minor thongs from me. Overhaul it looks already pretty good.

shyim committed:
opensearch: init at 2.5.0

shyim committed:
nixos/opensearch: init module

shyim committed:
nixos/tests/opensearch: init

shyim commented:
@mweinelt like so? copied the title like from https://github.com/NixOS/nixpkgs/commit/fe34d10e57b76696738b3a4a4352480d4e0ffd2f 😅

mweinelt commented on pkgs/servers/search/opensearch/default.nix:
```diff
@@ -0,0 +1,51 @@
+
+    runHook postInstall
+  '';
+
```
```suggestion
  passthru.tests = {
    inherit (nixosTests) opensearch;
  }:
```

This links the package to the nixos test, which will be run by ofborg when the package gets updated. Should go into the `nixos/tests/opensearch` commit.

shyim replied:
done :)

mweinelt commented:
Waiting for a second approval. Thanks for sinking your time into this.

mweinelt commented:
Oh, no! I forgot: New modules also require a changelog entry. See `nixos/doc/manual/release-notes/rl-2305.section.md` and the relevant checkboxes in the top post.

shyim committed:
nixos/doc: add release note for opensearch

shyim commented:
Updated release page, hope fully done. 😅

drupol reviewed (commented):
Massive work, thanks!

talyz committed:
nixos/opensearch: Use DynamicUser and StateDirectory by default

...but still allow for setting `dataDir` to a custom path. This gets
rid of the use of the deprecated option PermissionsStartOnly. Also, add
the ability to customize user and group, since that could be useful
with a custom `dataDir`.

talyz reviewed (changes requested):
Looks good overall! However, I thought it would be nice to get rid of `PermissionsStartOnly` and have `DynamicUser` and `StateDirectory` setup working, so I've opened https://github.com/shyim/nixpkgs/pull/1. It also includes some other minor fixes to the documentation and the start script.

shyim commented:
@talyz it's merged. many many thanks! ❤️ .

Do I need to cleanup the commits or it's fine so?

talyz commented on nixos/modules/services/search/opensearch.nix:
```diff
@@ -0,0 +1,244 @@
+{
+
+  options.services.opensearch = {
+    enable = mkEnableOption "OpenSearch";
```
Oops, I broke the manual build with this. Sorry about that! Should be
```suggestion
    enable = mkEnableOption (lib.mdDoc "OpenSearch");
```

shyim replied:
looks now good :)

talyz commented:
No problem! I think it's fine to keep them as separate commits.

talyz commented:
Great work! Thanks!

drupol commented:
Massive!
